### PR TITLE
Dealt with consecutive new lines. Wrote countOccurences rather than impo...

### PR DESCRIPTION
**Pull request by Njanderson**
...rting a library.

### CODE REVIEW

File: src/main/java/org/sagebionetworks/web/client/widget/entity/MarkdownEditorWidgetViewImpl.java [hunk 1]
```diff
@@ -274,48 +267,25 @@ public void onClick(ClickEvent event) {
...
+	public void addTextAreaKeyUpHandler(KeyUpHandler handler) {
+		markdownTextArea.addKeyUpHandler(handler);
+	}
+	
+	public void addTextAreaClickHandler(ClickHandler handler) {
+		markdownTextArea.addClickHandler(handler);
 	}
 	
-	private ClickHandler getDeleteClickHandler() {
+	private ClickHandler getClickHandler(final MarkdownEditorAction action) {
```
Comment (Njanderson) [inline src/main/java/org/sagebionetworks/web/client/widget/entity/MarkdownEditorWidgetViewImpl.java:295]: getClickHandler only includes: 

```
private ClickHandler getClickHandler(final MarkdownEditorAction action) {
    return new ClickHandler() {
        @Override
        public void onClick(ClickEvent event) {
            presenter.handleCommand(action);        
        }
    };
}
```

And I moved the construction of the delete button click handler to the presenter.

File: src/main/java/org/sagebionetworks/web/client/widget/entity/MarkdownEditorWidget.java [hunk 2]
```diff
@@ -148,9 +154,65 @@ public void configure(WikiPage page) {
...
+				markdownEditorClicked();
+			}
+		});
+	}
+	
+	private ClickHandler getDeleteClickHandler() {
+		return new ClickHandler() {
+			@Override
+			public void onClick(ClickEvent event) {
+				Bootbox.confirm(DisplayConstants.PROMPT_SURE_DELETE + " Page and Subpages?", new ConfirmCallback() {
```
Comment (jay-hodgson) [inline src/main/java/org/sagebionetworks/web/client/widget/entity/MarkdownEditorWidget.java]: Need to do more if we want to test this, as we can't refer to bootbox (ui element) in a presenter test (simply moving this code to the presenter does not make it well tested).  could ask the view to display bootbox and hand it the confirm callback.  Then a couple of tests should be added to exercise this (where confirm is true, and another with confirm as false)

File: src/main/java/org/sagebionetworks/web/client/widget/entity/MarkdownEditorWidget.java [hunk 3]
```diff
@@ -148,9 +154,65 @@ public void configure(WikiPage page) {
...
+						if (isConfirmed)
+							handleCommand(MarkdownEditorAction.DELETE);
+					}
+				});
+			}
+		};
+	}
+	
+	private void resizeMarkdownTextArea() {
+		if (view.getMarkdownTextAreaVisibleLines() < 5) {
```
Comment (jay-hodgson) [inline src/main/java/org/sagebionetworks/web/client/widget/entity/MarkdownEditorWidget.java]: tests for resizeMarkdownTextArea() (you can make public to expose for test)

File: src/main/java/org/sagebionetworks/web/client/widget/entity/MarkdownEditorWidget.java [hunk 4]
```diff
@@ -148,7 +161,63 @@ public void configure(WikiPage page) {
...
+	public void resizeMarkdownTextArea() {
+		int visLines = view.getMarkdownTextAreaVisibleLines();
+		int index = 0;
+		int numLines = 0;
+		String editorText = view.getMarkdownText();
+		do {
+			index = 1 + editorText.indexOf("\n",index);
+			numLines++;
+		} while (index > 0 && index < editorText.length());
+		if (visLines < 5 || visLines != numLines + 1) {
```
Comment (jay-hodgson) [inline src/main/java/org/sagebionetworks/web/client/widget/entity/MarkdownEditorWidget.java]: pull 5 out as a constant (MIN_LINES)

File: src/main/java/org/sagebionetworks/web/client/widget/entity/MarkdownEditorWidgetViewImpl.java [hunk 5]
```diff
@@ -274,48 +272,33 @@ public void onClick(ClickEvent event) {
...
 	
-	private ClickHandler getClickHandler(final MarkdownEditorAction action) {
-		return new ClickHandler() {
-			@Override
-			public void onClick(ClickEvent event) {
-				presenter.handleCommand(action);		
-			}
-		};
+	@Override 
+	public void confirmDeletion(String text, ConfirmCallback callback) {
```
Comment (jay-hodgson) [inline src/main/java/org/sagebionetworks/web/client/widget/entity/MarkdownEditorWidgetViewImpl.java]: rename to confirm()

File: src/test/java/org/sagebionetworks/web/unitclient/widget/entity/MarkdownEditorWidgetTest.java [hunk 6]
```diff
@@ -192,6 +198,30 @@ public void testPreviewFailure() throws Exception {
 	}
 	
 	@Test
+	public void testDeleteConfirmedCallback() {
+		ClickHandler deleteClickHandler = presenter.getDeleteClickHandler();
+		mockView.setDeleteClickHandler(deleteClickHandler);
```
Comment (jay-hodgson) [inline src/test/java/org/sagebionetworks/web/unitclient/widget/entity/MarkdownEditorWidgetTest.java]: remove, but check setDeleteClickHandler in configure()

File: src/test/java/org/sagebionetworks/web/unitclient/widget/entity/MarkdownEditorWidgetTest.java [hunk 7]
```diff
@@ -192,6 +198,30 @@ public void testPreviewFailure() throws Exception {
...
+	}
+	
+	@Test
+	public void testDeleteCancelCallback() {
+		ClickHandler deleteClickHandler = presenter.getDeleteClickHandler();
+		mockView.setDeleteClickHandler(deleteClickHandler);
+		deleteClickHandler.onClick(null);
+		ArgumentCaptor<ConfirmCallback> captor = ArgumentCaptor.forClass(ConfirmCallback.class);
+		verify(mockView).confirmDeletion(anyString(), captor.capture());
+		//confirm deletion
```
Comment (jay-hodgson) [inline src/test/java/org/sagebionetworks/web/unitclient/widget/entity/MarkdownEditorWidgetTest.java]: boolean isConfirmed=false;

File: src/test/java/org/sagebionetworks/web/unitclient/widget/entity/MarkdownEditorWidgetTest.java [hunk 8]
```diff
@@ -250,6 +280,86 @@ public void testInsertMarkdownEnd() {
 	}
 	
 	@Test
+	public void testResizeMarkdownEmpty() {
+		String markdown = "";
+		when(mockView.getMarkdownTextAreaVisibleLines()).thenReturn(0);
+		when(mockView.getMarkdownText()).thenReturn(markdown);
+		when(mockView.getCursorPos()).thenReturn(4);
+		presenter.resizeMarkdownTextArea();
+		verify(mockView).resizeMarkdownTextArea(5);
```
Comment (jay-hodgson) [inline src/test/java/org/sagebionetworks/web/unitclient/widget/entity/MarkdownEditorWidgetTest.java]: refer to min lines constant

**Pull request merged**